renku/api/datasets.py

```diff
@@ -282,141 +284,166 @@ def _add_from_url(self, dataset, path, url, link=False, **kwargs):
-        relative_to = kwargs.get('relative_to', None)
 
-        if u.scheme in ('', 'file'):
+        if u.scheme in ('', 'file') and not url.startswith('git@'):
             try:
-                relative_url = Path(url).resolve().relative_to(self.path)
-            except Exception:
-                relative_url = None
-
-            if relative_url:
-                return [{
-                    'path': url,
-                    'url': url,
-                    'creator': dataset.creator,
-                    'dataset': dataset.name,
-                    'parent': self
-                }]
+                Path(u.path).resolve().relative_to(self.path)
+            except ValueError:
+                pass
+            else:
+                if not destination:
+                    return [{
+                        'path': url,
+                        'url': url,
+                        'creator': dataset.creator,
+                        'dataset': dataset.name,
+                        'parent': self
+                    }]
 
             warnings.warn('Importing local git repository, use HTTPS')
+
             # determine where is the base repo path
-            r = Repo(url, search_parent_directories=True)
-            src_repo_path = Path(r.git_dir).parent.resolve()
-            submodule_name = src_repo_path.name
-            submodule_path = submodule_path / str(src_repo_path).lstrip('/')
-
-            # if repo path is a parent, rebase the paths and update url
-            if src_repo_path != Path(u.path):
-                top_target = Path(
-                    u.path
-                ).resolve().absolute().relative_to(src_repo_path)
-                if target:
-                    target = top_target / target
-                else:
-                    target = top_target
-                url = src_repo_path.as_posix()
-        elif u.scheme in {'http', 'https', 'git+https', 'git+ssh'}:
-            submodule_name = os.path.splitext(os.path.basename(u.path))[0]
-            submodule_path = submodule_path.joinpath(
-                os.path.dirname(u.path).lstrip('/'), submodule_name
-            )
+            repo = Repo(url, search_parent_directories=True)
+            repo_path = Path(repo.git_dir).parent.resolve()
+
+            # if repo path is a parent of the url, treat the url as a source
+            if repo_path != Path(u.path):
+                if sources:
+                    raise errors.UsageError(
+                        'Cannot use --source within local repo subdirectories'
+                    )
+                source = Path(u.path).resolve().relative_to(repo_path)
+                sources = (source, )
+                url = repo_path.as_posix()
+        elif u.scheme in {'http', 'https', 'git+https', 'git+ssh'} or \
+                url.startswith('git@'):
+            repo_path = Path(tempfile.mkdtemp())
+            repo = Repo.clone_from(url, repo_path, recursive=True)
```

Comment:
what happens with LFS objects here? We may want to use the `GIT_LFS_SKIP_SMUDGE=1` environment variable to prevent all the data being pulled and then only fetch the needed paths.